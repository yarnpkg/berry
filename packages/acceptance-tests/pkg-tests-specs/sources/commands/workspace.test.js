const {
  fs: {writeFile, writeJson},
} = require(`pkg-tests-core`);


async function setupWorkspaces(path) {
  await writeFile(`${path}/.yarnrc.yml`, `plugins:\n  - ${JSON.stringify(require.resolve(`@yarnpkg/monorepo/scripts/plugin-workspace-tools.js`))}\n`);

  await writeJson(`${path}/docs/package.json`, {
    name: `docs`,
    version: `1.0.0`,
    scripts: {
      print: `echo Test Docs`,
    },
    dependencies: {
      [`component-a`]: `workspace:*`,
    },
  });

  await writeJson(`${path}/components/component-a/package.json`, {
    name: `component-a`,
    version: `1.0.0`,
    scripts: {
      print: `echo Test Component A`,
    },
    dependencies: {
      [`component-b`]: `workspace:*`,
    },
  });

  await writeJson(`${path}/components/component-b/package.json`, {
    name: `component-b`,
    version: `1.0.0`,
    scripts: {
      print: `echo Test Component B`,
    },
  });
}

describe(`Commands`, () => {
  describe(`workspace <workspace-name> <sub-command>`, () => {
    test(
      `runs a given command in the specified workspace`,
      makeTemporaryEnv(
        {
          private: true,
          workspaces: [`docs`, `components/*`],
        },
        async ({path, run}) => {
          await setupWorkspaces(path);

          let code;
          let stdout;
          let stderr;

          try {
            await run(`install`);
            ({code, stdout, stderr} = await run(`workspace`, `component-a`, `print`));
          } catch (error) {
            ({code, stdout, stderr} = error);
          }

          await expect({code, stdout, stderr}).toMatchSnapshot();
        }
      )
    );

    test(
      `when the given workspace doesnt exist lists all possible workspaces`,
      makeTemporaryEnv(
        {
          private: true,
          workspaces: [`docs`, `components/*`],
        },
        async ({path, run}) => {
          await setupWorkspaces(path);

          let code;
          let stdout;
          let stderr;

          try {
            await run(`install`);
            ({code, stdout, stderr} = await run(`workspace`, `compoment-a`, `print`));
          } catch (error) {
            ({code, stdout, stderr} = error);
          }

          await expect({code, stdout, stderr}).toMatchSnapshot();
        }
      )
    );
  });
});
