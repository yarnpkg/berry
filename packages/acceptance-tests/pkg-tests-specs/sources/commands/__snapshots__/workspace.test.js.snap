// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Commands workspace <workspace-name> <sub-command> runs a given command in the specified workspace 1`] = `
Object {
  "code": 0,
  "stderr": "",
  "stdout": "Test Component A
",
}
`;

exports[`Commands workspace <workspace-name> <sub-command> when the given workspace doesnt exist lists all possible workspaces 1`] = `
Object {
  "code": 1,
  "stderr": "",
  "stdout": "Usage Error: Workspace 'compoment-a' not found. Did you mean any of the following:
  - component-a
  - component-b
  - docs
  - root-workspace-988eec?

$ yarn workspace <workspaceName> <commandName> ...
",
}
`;
